core/src/models/ExportProcessor.ts

```diff
@@ -0,0 +1,225 @@
+
+  @Column
+  errorMessage: string;
+
+  @HasMany(() => Export)
+  exports: Export[];
+
+  @Is("ofValidErrorLevel", (value) => {
+    if (value && !ERROR_LEVELS.includes(value)) {
+      throw new Error(`errorLevel must be one of: ${ERROR_LEVELS.join(",")}`);
+    }
+  })
+  @Column(DataType.ENUM(...ERROR_LEVELS))
+  errorLevel: ExportProcessorErrorLevel;
+
+  @BelongsTo(() => Destination)
+  destination: Destination;
+
+  async setError(error: Error, retryDelay: number = config.tasks.timeout) {
+    return ExportProcessorOps.setError(this, error, retryDelay);
+  }
+
+  async retry(retryDelay: number = config.tasks.timeout) {
+    return ExportProcessorOps.retry(this, retryDelay);
+  }
+
+  async complete() {
+    this.errorMessage = null;
+    this.errorLevel = null;
+    this.completedAt = new Date();
+    this.state = "complete";
+    await this.save();
+  }
+
+  async apiData(includeDestination = true) {
+    const exportsCount = await this.$count("exports");
+    const destination = await this.$get("destination", {
+      scope: null,
+    });
+
+    return {
+      id: this.id,
+      destination:
+        destination && includeDestination
+          ? await destination.apiData(false, false)
+          : null,
+      destinationName: destination ? destination.name : null,
+      state: this.state,
+      createdAt: APIData.formatDate(this.createdAt),
+      processAt: APIData.formatDate(this.processAt),
+      startedAt: APIData.formatDate(this.startedAt),
+      completedAt: APIData.formatDate(this.completedAt),
+      remoteKey: this.remoteKey,
+      exportsCount,
+      retryCount: this.retryCount,
+      errorMessage: this.errorMessage,
+      errorLevel: this.errorLevel,
+    };
+  }
+
+  async getExportsToProcess() {
+    return ExportProcessorOps.getExportsToProcess(this);
+  }
+
+  async ensureAllExportsProcessed() {
+    const pendingExportsCount = await this.$count("exports", {
+      where: { state: { [Op.notIn]: ["complete", "failed"] } },
+    });
+
+    if (pendingExportsCount > 0) {
+      throw new Error(
+        "Cannot mark export processor as complete because some exports have not been processed"
+      );
+    }
+  }
+
+  // --- Class Methods --- //
+
+  static defaultState = "pending";
+
+  static async findById(id: string) {
+    const instance = await this.scope(null).findOne({ where: { id } });
+    if (!instance) throw new Error(`cannot find ${this.name} ${id}`);
+    return instance;
+  }
+
+  @BeforeSave
+  static async updateState(instance: ExportProcessor) {
+    await StateMachine.transition(instance, STATE_TRANSITIONS);
+  }
+
+  @BeforeCreate
+  static generateId(instance) {
+    if (!instance.id) {
+      instance.id = `${instance.idPrefix()}_${uuid.v4()}`;
+    }
+  }
+
+  @BeforeSave
+  static ensureErrorLevel(instance: ExportProcessor) {
+    if (instance.errorMessage && !instance.errorLevel) {
+      instance.errorLevel = "error";
+    }
+  }
+
+  static async sweep(limit: number) {
```

Comment:
👍 

Note to self: The reason to not sweep an ExportProcessor that has exports is to maintain any foreign key relationships.  Exports are already being swept every N days, so that functionally makes ExportProcessors sept on N+1 days.  Neat.